File: packages/one/android/src/main/java/com/margelo/nitro/one/HybridOneFetch.kt
package com.margelo.nitro.one

import android.net.Uri
import android.webkit.MimeTypeMap
import com.facebook.react.modules.blob.BlobModule
import com.facebook.react.modules.network.CookieJarContainer
import com.facebook.react.modules.network.ForwardingCookieHandler
import com.facebook.react.modules.network.OkHttpClientProvider
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.ArrayBuffer
import java.io.IOException
import java.io.InputStream
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.CookieJar
import okhttp3.Headers
import okhttp3.JavaNetCookieJar
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.internal.http.HttpMethod
import okio.BufferedSink

// the global fetch on native, over an OkHttp client built the way react
// native's NetworkingModule builds its own: from OkHttpClientProvider, so an
// app's OkHttpClientFactory (interceptors, certificate pinning) applies, with
// its cookie jar backed by the same CookieManager store react native uses. the response head goes to js first, then every chunk as the
// socket delivers it, so response.body streams. file: and content: urls are
// read from disk, as react native's fetch reads them.
class HybridOneFetch : HybridOneFetchSpec() {
    // cancel actions for live requests; whoever removes an id owns its
    // terminal event, so a cancelled request never reports
    private val active = ConcurrentHashMap<Double, () -> Unit>()
    private val local = Executors.newCachedThreadPool()

    private val client: OkHttpClient by lazy {
        OkHttpClientProvider.createClient().also {
            (it.cookieJar as? CookieJarContainer)
                ?.setCookieJar(JavaNetCookieJar(ForwardingCookieHandler()))
        }
    }
    private val clientWithoutCookies: OkHttpClient by lazy {
        client.newBuilder().cookieJar(CookieJar.NO_COOKIES).build()
    }

    private val context
        get() = NitroModules.applicationContext
            ?: throw IllegalStateException("fetch: the react context is not ready")

    override fun start(
        id: Double,
        request: FetchNativeRequest,
        onResponse: (response: FetchNativeResponse) -> Unit,
        onChunk: (chunk: ArrayBuffer) -> Unit,
        onComplete: () -> Unit,
        onError: (message: String) -> Unit
    ) {
        val scheme = Uri.parse(request.url).scheme
        if (scheme == "file" || scheme == "content") {
            readLocal(id, request.url, onResponse, onChunk, onComplete, onError)
            return
        }
        val built = try {
            val builder = Request.Builder().url(request.url)
            for (header in request.headers) builder.addHeader(header.name, header.value)
            val type = request.headers
                .lastOrNull { it.name.equals("content-type", ignoreCase = true) }
                ?.value?.toMediaTypeOrNull()
            builder.method(request.method, body(request, type))
            builder.build()
        } catch (error: IllegalArgumentException) {
            onError(error.message ?: "invalid request")
            return
        }
        val call = (if (request.omitCredentials) clientWithoutCookies else client).newCall(built)
        active[id] = { call.cancel() }
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (active.remove(id) != null) onError(e.message ?: e.javaClass.simpleName)
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (!active.containsKey(id)) return
                    val headers = response.headers.map { (name, value) -> FetchHeader(name, value) }
                    onResponse(
                        FetchNativeResponse(
                            response.code.toDouble(),
                            response.message,
                            response.request.url.toString(),
                            response.priorResponse != null,
                            headers.toTypedArray()
                        )
                    )
                    pump(id, response.body?.byteStream(), onChunk, onComplete, onError)
                }
            }
        })
    }

    override fun cancel(id: Double) {
        active.remove(id)?.invoke()
    }

    override fun storeBlob(bytes: ArrayBuffer): String = blobModule().store(bytes.toByteArray())

    private fun blobModule(): BlobModule =
        context.getNativeModule(BlobModule::class.java)
            ?: throw IllegalStateException("fetch: react native's BlobModule is not loaded")

    // the request body. js-owned bytes are copied now, during the call;
    // blobs, files and multipart parts are read on the network thread. every
    // length is known up front where react native's was, so uploads carry a
    // content-length rather than going out chunked.
    private fun body(request: FetchNativeRequest, type: MediaType?): RequestBody? {
        request.body?.let { return it.toByteArray().toRequestBody(type) }
        request.blob?.let { return blobBody(it, type) }
        val form = request.form
        val boundary = request.boundary
        if (form != null && boundary != null) return multipart(form, boundary)
        return if (HttpMethod.requiresRequestBody(request.method)) ByteArray(0).toRequestBody(type)
        else null
    }

    private fun blobBody(blob: FetchBlobRef, type: MediaType?) = object : RequestBody() {
        override fun contentType() = type
        override fun contentLength() = blob.size.toLong()
        override fun writeTo(sink: BufferedSink) {
            sink.write(resolve(blob))
        }
    }

    private fun uriBody(uri: String, type: MediaType?) = object : RequestBody() {
        override fun contentType() = type
        override fun contentLength() =
            context.contentResolver.openAssetFileDescriptor(Uri.parse(uri), "r")?.use { it.length }
                ?: -1L
        override fun writeTo(sink: BufferedSink) {
            open(uri).use { input -> input.copyTo(sink.outputStream()) }
        }
    }

    private fun resolve(blob: FetchBlobRef): ByteArray =
        blobModule().resolve(blob.blobId, blob.offset.toInt(), blob.size.toInt())
            ?: throw IOException("fetch: blob ${blob.blobId} is no longer in the store")

    private fun open(uri: String): InputStream =
        context.contentResolver.openInputStream(Uri.parse(uri))
            ?: throw IOException("fetch: could not open $uri")

    // multipart/form-data per RFC 7578 through okhttp's MultipartBody, as
    // react native builds it, under the boundary js already put in the
    // content-type header
    private fun multipart(parts: Array<FetchFormPart>, boundary: String): RequestBody {
        val builder = MultipartBody.Builder(boundary).setType(MultipartBody.FORM)
        for (part in parts) {
            var disposition = "form-data; name=\"${escape(part.name)}\""
            val value = part.value
            if (value != null) {
                builder.addPart(Headers.headersOf("Content-Disposition", disposition), value.toRequestBody())
                continue
            }
            val filename = part.filename ?: part.uri?.let { Uri.parse(it).lastPathSegment }
            if (filename != null) disposition += "; filename=\"${escape(filename)}\""
            val type = (part.type ?: "application/octet-stream").toMediaTypeOrNull()
            val uri = part.uri
            val blob = part.blob
            val body = when {
                uri != null -> uriBody(uri, type)
                blob != null -> blobBody(blob, type)
                else -> throw IOException("fetch: FormData part ${part.name} has no value")
            }
            builder.addPart(Headers.headersOf("Content-Disposition", disposition), body)
        }
        return builder.build()
    }

    private fun escape(value: String) =
        value.replace("\"", "%22").replace("\r", "%0D").replace("\n", "%0A")

    private fun readLocal(
        id: Double,
        url: String,
        onResponse: (response: FetchNativeResponse) -> Unit,
        onChunk: (chunk: ArrayBuffer) -> Unit,
        onComplete: () -> Unit,
        onError: (message: String) -> Unit
    ) {
        active[id] = {}
        local.execute {
            val input = try {
                open(url)
            } catch (error: Exception) {
                if (active.remove(id) != null) onError(error.message ?: "could not open $url")
                return@execute
            }
            input.use {
                val uri = Uri.parse(url)
                val type = context.contentResolver.getType(uri)
                    ?: MimeTypeMap.getSingleton().getMimeTypeFromExtension(
                        MimeTypeMap.getFileExtensionFromUrl(url)
                    )
                val headers = type?.let { arrayOf(FetchHeader("content-type", it)) } ?: emptyArray()
                if (!active.containsKey(id)) return@execute
                onResponse(FetchNativeResponse(200.0, "", url, false, headers))
                pump(id, input, onChunk, onComplete, onError)
            }
        }
    }

    private fun pump(
        id: Double,
        input: InputStream?,
        onChunk: (chunk: ArrayBuffer) -> Unit,
        onComplete: () -> Unit,
        onError: (message: String) -> Unit
    ) {
        val buffer = ByteArray(CHUNK)
        try {
            while (input != null) {
                val read = input.read(buffer)
                if (read == -1) break
                if (!active.containsKey(id)) return
                if (read > 0) onChunk(ArrayBuffer.copy(buffer.copyOf(read)))
            }
        } catch (error: IOException) {
            if (active.remove(id) != null) onError(error.message ?: error.javaClass.simpleName)
            return
        }
        if (active.remove(id) != null) onComplete()
    }

    override fun dispose() {
        active.values.forEach { it() }
        active.clear()
        local.shutdownNow()
        super.dispose()
    }

    companion object {
        private const val CHUNK = 16 * 1024
    }
}

File: packages/one/ios/Nitro/HybridOneFetch.swift
import Foundation
import NitroModules

// the global fetch on native, over URLSession. the session delegate hands the
// response head, then every chunk as it arrives, to js, so response.body
// streams. the configuration matches react native's RCTHTTPRequestHandler
// (shared cookie storage, cookies always accepted), so this fetch and
// react native networking see the same cookies. URLSession also reads file:
// and data: urls, as react native's fetch does.
final class HybridOneFetch: HybridOneFetchSpec {
  private final class Request {
    let onResponse: (FetchNativeResponse) -> Void
    let onChunk: (ArrayBuffer) -> Void
    let onComplete: () -> Void
    let onError: (String) -> Void
    var redirected = false

    init(
      onResponse: @escaping (FetchNativeResponse) -> Void,
      onChunk: @escaping (ArrayBuffer) -> Void,
      onComplete: @escaping () -> Void,
      onError: @escaping (String) -> Void
    ) {
      self.onResponse = onResponse
      self.onChunk = onChunk
      self.onComplete = onComplete
      self.onError = onError
    }
  }

  // all request state lives on the session's serial delegate queue
  private final class Delegate: NSObject, URLSessionDataDelegate {
    let queue: OperationQueue = {
      let queue = OperationQueue()
      queue.maxConcurrentOperationCount = 1
      queue.name = "one.fetch"
      return queue
    }()
    var requests: [Int: Request] = [:]
    var tasks: [Double: URLSessionTask] = [:]

    func urlSession(
      _ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse,
      completionHandler: @escaping (URLSession.ResponseDisposition) -> Void
    ) {
      guard let request = requests[dataTask.taskIdentifier] else {
        completionHandler(.cancel)
        return
      }
      // file: and data: responses are not http; they read as a plain 200
      let http = response as? HTTPURLResponse
      var headers: [FetchHeader] = []
      if let http {
        for (name, value) in http.allHeaderFields {
          headers.append(FetchHeader(name: "\(name)", value: "\(value)"))
        }
      } else if let type = response.mimeType {
        headers.append(FetchHeader(name: "content-type", value: type))
      }
      // URLSession exposes no reason phrase; http/2 has none either
      request.onResponse(
        FetchNativeResponse(
          status: Double(http?.statusCode ?? 200),
          statusText: "",
          url: response.url?.absoluteString ?? dataTask.originalRequest?.url?.absoluteString ?? "",
          redirected: request.redirected,
          headers: headers))
      completionHandler(.allow)
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
      guard let request = requests[dataTask.taskIdentifier] else { return }
      do {
        request.onChunk(try ArrayBuffer.copy(data: data))
      } catch {
        fail(dataTask, error.localizedDescription)
      }
    }

    func urlSession(
      _ session: URLSession, task: URLSessionTask,
      willPerformHTTPRedirection response: HTTPURLResponse, newRequest: URLRequest,
      completionHandler: @escaping (URLRequest?) -> Void
    ) {
      requests[task.taskIdentifier]?.redirected = true
      completionHandler(newRequest)
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
      guard let request = forget(task) else { return }
      if let error {
        request.onError(error.localizedDescription)
      } else {
        request.onComplete()
      }
    }

    func fail(_ task: URLSessionTask, _ message: String) {
      guard let request = forget(task) else { return }
      task.cancel()
      request.onError(message)
    }

    @discardableResult
    func forget(_ task: URLSessionTask) -> Request? {
      if let id = tasks.first(where: { $0.value === task })?.key {
        tasks[id] = nil
      }
      return requests.removeValue(forKey: task.taskIdentifier)
    }
  }

  private let delegate = Delegate()
  private let session: URLSession

  override init() {
    let configuration = URLSessionConfiguration.default
    configuration.httpShouldSetCookies = true
    configuration.httpCookieAcceptPolicy = .always
    configuration.httpCookieStorage = .shared
    session = URLSession(
      configuration: configuration, delegate: delegate, delegateQueue: delegate.queue)
    super.init()
  }

  deinit {
    session.invalidateAndCancel()
  }

  func start(
    id: Double, request: FetchNativeRequest,
    onResponse: @escaping (FetchNativeResponse) -> Void,
    onChunk: @escaping (ArrayBuffer) -> Void,
    onComplete: @escaping () -> Void,
    onError: @escaping (String) -> Void
  ) throws {
    guard let url = URL(string: request.url) else {
      onError("invalid url \(request.url)")
      return
    }
    var urlRequest = URLRequest(url: url)
    urlRequest.httpMethod = request.method
    urlRequest.httpShouldHandleCookies = !request.omitCredentials
    for header in request.headers {
      urlRequest.addValue(header.value, forHTTPHeaderField: header.name)
    }
    // js-owned buffers are only readable during this call, so copy now
    let body = request.body.map { $0.toData(copyIfNeeded: true) }
    let blob = request.blob
    let form = request.form
    let boundary = request.boundary
    let pending = Request(
      onResponse: onResponse, onChunk: onChunk, onComplete: onComplete, onError: onError)
    delegate.queue.addOperation { [delegate, session] in
      do {
        if let body {
          urlRequest.httpBody = body
        } else if let blob {
          urlRequest.httpBody = try Self.resolve(blob)
        } else if let form, let boundary {
          urlRequest.httpBody = try Self.multipart(form, boundary: boundary)
        }
      } catch {
        onError(error.localizedDescription)
        return
      }
      let task = session.dataTask(with: urlRequest)
      delegate.requests[task.taskIdentifier] = pending
      delegate.tasks[id] = task
      task.resume()
    }
  }

  func cancel(id: Double) throws {
    delegate.queue.addOperation { [delegate] in
      guard let task = delegate.tasks.removeValue(forKey: id) else { return }
      delegate.requests[task.taskIdentifier] = nil
      task.cancel()
    }
  }

  func storeBlob(bytes: ArrayBuffer) throws -> String {
    let data = bytes.toData(copyIfNeeded: true)
    guard let blobId = try Self.blobManager().perform(NSSelectorFromString("store:"), with: data)?
      .takeUnretainedValue() as? String
    else {
      throw RuntimeError.error(withMessage: "fetch: the react native blob store rejected the bytes")
    }
    return blobId
  }

  // react native's blob store, the BlobModule its networking and Blob use
  private static func blobManager() throws -> NSObject {
    guard let manager = OneFetchBlobStore.manager else {
      throw RuntimeError.error(withMessage: "fetch: react native's BlobModule is not loaded")
    }
    return manager
  }

  private static func resolve(_ blob: FetchBlobRef) throws -> Data {
    let ref: NSDictionary = [
      "blobId": blob.blobId, "offset": Int(blob.offset), "size": Int(blob.size),
    ]
    guard let data = try blobManager().perform(NSSelectorFromString("resolve:"), with: ref)?
      .takeUnretainedValue() as? Data
    else {
      throw RuntimeError.error(withMessage: "fetch: blob \(blob.blobId) is no longer in the store")
    }
    return data
  }

  private static func read(uri: String) throws -> Data {
    guard let url = URL(string: uri), url.isFileURL else {
      throw RuntimeError.error(withMessage: "fetch: FormData uri \(uri) is not a file: url")
    }
    return try Data(contentsOf: url)
  }

  // multipart/form-data per RFC 7578, the same layout react native writes
  private static func multipart(_ parts: [FetchFormPart], boundary: String) throws -> Data {
    var body = Data()
    func line(_ text: String) { body.append(Data((text + "\r\n").utf8)) }
    for part in parts {
      line("--\(boundary)")
      var disposition = "Content-Disposition: form-data; name=\"\(escape(part.name))\""
      if let filename = part.filename ?? part.uri.flatMap({ URL(string: $0)?.lastPathComponent }),
        part.value == nil
      {
        disposition += "; filename=\"\(escape(filename))\""
      }
      line(disposition)
      if let value = part.value {
        line("")
        line(value)
        continue
      }
      let data: Data
      if let uri = part.uri {
        data = try read(uri: uri)
      } else if let blob = part.blob {
        data = try resolve(blob)
      } else {
        throw RuntimeError.error(withMessage: "fetch: FormData part \(part.name) has no value")
      }
      line("Content-Type: \(part.type ?? "application/octet-stream")")
      line("")
      body.append(data)
      line("")
    }
    line("--\(boundary)--")
    return body
  }

  private static func escape(_ value: String) -> String {
    value.replacingOccurrences(of: "\"", with: "%22").replacingOccurrences(of: "\r", with: "%0D")
      .replacingOccurrences(of: "\n", with: "%0A")
  }
}

// set by OneFetchBlobStoreModule from react native's module registry
@objcMembers public final class OneFetchBlobStore: NSObject {
  public static weak var manager: NSObject?
}
